### DISCUSSION TITLE
"docker: not found" on run step

### DISCUSSION THREAD

**Discussion by AssisrMatheus** [community/community#25268] [Actions]
<p>I am running a workflow on each PR to test my application with cypress. This workflow raises two <a href="https://docs.github.com/en/actions/using-containerized-services/about-service-containers" rel="noopener nofollow ugc">service containers</a>, one for the database, and another for the backend,</p>
<p>Then, in one of the steps. I try to run the seed inside the backend service. To get the database ready for testing.</p>
<p>But unfortunately this step is failing with the following message:</p>
<pre><code class="lang-auto">Run docker exec d1c0b533b80d0eee96631d4024cd3c2bc5bb710f950949f747edb4a4b304f516 sh -c 'npm run seed'
/__w/_temp/cb64157a-9be2-4ea5-83b9-20d8a79af6bf.sh: 1: /__w/_temp/cb64157a-9be2-4ea5-83b9-20d8a79af6bf.sh: docker: not found
Error: Process completed with exit code 127.
</code></pre>
<p>I also tried using the <code>--entrypoint</code> option in the backend service, having it run an specific script in <code>package.json</code> that seeds <strong>and</strong> then starts the server, but then I get another error:</p>
<pre><code class="lang-auto">/usr/bin/docker start 49fef80427d4bae51d14f00eeea4b1039ce3c11037fe6ef1afebc37f9f9cc498
  Error response from daemon: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "npm run test:frontend": executable file not found in $PATH: unknown
  Error: failed to start containers: 49fef80427d4bae51d14f00eeea4b1039ce3c11037fe6ef1afebc37f9f9cc498
  Error: Docker start fail with exit code 1
</code></pre>
<p>I’m now out of options on how to start a service and have my tests depend on it. I had it working when I started the services as a step, using <code>docker-compose</code>(which now also doesn’t work).</p>
<p>But when I use the <code>container: cypress/browsers:node14.15.0-chrome96-ff94</code> on my job, since I depend on having chrome and firefox for tests, and this image, officially made by cypress has it, docker does not work anymore.</p>
<p>I assume that is the reason of the problem. But I don’t know the reason why.</p>
<p>My workflow file is similar to this, notice the “run seed” step:</p>
<pre><code class="lang-yaml">name: Continuous Integration

on:
  workflow_dispatch:
  pull_request:
    branches: [dev]

jobs:
  install-and-build:
    timeout-minutes: 10
    runs-on: ubuntu-latest
    container: cypress/browsers:node14.15.0-chrome96-ff94
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          registry-url: 'https://npm.pkg.github.com'
      - name: Build and cache for cypress
        uses: cypress-io/github-action@v2
        with:
          runTests: false
          build: npm run build
      - name: Save build folder
        uses: actions/upload-artifact@v2
        with:
          name: .next
          if-no-files-found: error
          path: .next

  test-chrome:
    timeout-minutes: 15
    runs-on: ubuntu-latest
    container: cypress/browsers:node14.15.0-chrome96-ff94
    needs: install-and-build
    services:
      db:
        image: mysql
        ports:
          - 3306:3306
        env:
          MYSQL_ROOT_PASSWORD: root
        options: &gt;-
          --health-cmd "mysqladmin ping -h localhost --silent --wait=20"
          --health-interval 30s
          --health-timeout 25s
          --health-retries 10
          --health-start-period 5s
      backend:
        image: 'path/to/prebuilt/backend/image'
        ports:
          - 3001:3001
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          registry-url: 'https://npm.pkg.github.com'
      - name: Run seed
        run: |
          docker exec ${{ job.services.backend.id }} sh -c 'npm run seed'
          docker kill --signal=SIGHUP ${{ job.services.backend.id }}
      - name: Download the build folders
        uses: actions/download-artifact@v2
        with:
          name: .next
          path: .next
      - name: Run cypress with chrome
        uses: cypress-io/github-action@v2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          start: npm run test:serve
          wait-on: 'http://localhost:3001, http://localhost:3000'
          wait-on-timeout: 120
          browser: chrome
      - uses: actions/upload-artifact@v2
        if: failure()
        with:
          name: cypress-screenshots-chrome
          path: cypress/screenshots
      - uses: actions/upload-artifact@v2
        if: always()
        with:
          name: cypress-videos-chrome
          path: cypress/videos
</code></pre>
<p>I also tried changing the <code>docker</code> command to <code>/usr/bin/docker</code>, but it yields the same result</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="AssisrMatheus" data-post="1" data-topic="232003">
<div class="title">
<div class="quote-controls"></div>
 AssisrMatheus:</div>
<blockquote>
<p>But when I use the <code>container: cypress/browsers:node14.15.0-chrome96-ff94</code> on my job, since I depend on having chrome and firefox for tests, and this image, officially made by cypress has it, docker does not work anymore.</p>
</blockquote>
</aside>
<p>Does that image come with Docker installed? The Actions runner mounts the Docker socket into the container by default (so you can use the Docker service running on the runner VM), but the Docker binary still needs to be installed in the image.</p>

**Comment by AssisrMatheus** [comment]
<p>Not sure but would assume not then. But does that mean I have to install docker <em><strong>within docker</strong></em> ?</p>
<p>How would you suggest doing that within the dockerfile?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="AssisrMatheus" data-post="3" data-topic="232003">
<div class="title">
<div class="quote-controls"></div>
 AssisrMatheus:</div>
<blockquote>
<p>But does that mean I have to install docker <em><strong>within docker</strong></em> ?</p>
</blockquote>
</aside>
<p>Partially. The <code>docker</code> command talks to a service (<code>dockerd</code>) over a socket, and <code>dockerd</code> then runs the actual containers. The Actions runner makes sure that socket is available inside a job container, but if you actually want to talk to it you need to have something that speaks the language. So you either need the <code>docker</code> binary (or roll your own).</p>
<aside class="quote no-group" data-username="AssisrMatheus" data-post="3" data-topic="232003">
<div class="title">
<div class="quote-controls"></div>
 AssisrMatheus:</div>
<blockquote>
<p>How would you suggest doing that within the dockerfile?</p>
</blockquote>
</aside>
<p>Depends on what the image is based on. Most start <code>FROM</code> the base image of a Linux distribution, often Debian or Alpine. Preferably figure out which, and use its package manager to install the appropriate package.</p>

**Comment by AssisrMatheus** [comment]
<p>I’d assume my <code>FROM</code> would be <code>cypress/browsers:node14.15.0-chrome96-ff94</code> and then I’d have to install docker in it some way, to have access to the binary.</p>
<p>I’ll try that after the weekend and follow up with the results</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="AssisrMatheus" data-post="5" data-topic="232003">
<div class="title">
<div class="quote-controls"></div>
 AssisrMatheus:</div>
<blockquote>
<p>I’d assume my <code>FROM</code> would be <code>cypress/browsers:node14.15.0-chrome96-ff94</code> and then I’d have to install docker in it some way, to have access to the binary.</p>
</blockquote>
</aside>
<p>Yes. My point was to look at the <code>FROM</code> line of the Dockerfile for that image (and possibly its base in turn), to figure out which distribution it is based on. Then using that distribution’s package manager to install Docker would be much simpler than figuring out the commands to do it from scratch.</p>

**Comment by AssisrMatheus** [comment]
<p>After some trial and error I managed to get docker installed with these steps:</p>
<pre><code class="lang-auto">      - name: Install docker
        run: |
          apt-get update &amp;&amp; apt-get install --no-install-recommends -y ca-certificates curl gnupg lsb-release sudo
          curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
          echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list &gt; /dev/null
          apt-get update &amp;&amp; apt-get install --no-install-recommends -y docker-ce docker-ce-cli containerd.io

      - name: Run seed
        run: |
          docker exec ${{ job.services.backend.id }} sh -c 'npm run seed'
          docker kill --signal=SIGHUP ${{ job.services.backend.id }}
</code></pre>
<p>And docker now works within the image, but now a new issue appears.</p>
<pre><code class="lang-auto">Error response from daemon: Container fb0ba7f45e14c22342cf1c80662864cb432fbe430f2cee7997ee3ba211875fa3 is not running
</code></pre>
<p>This is probably due to the service containers not actually running within the same instance that the steps runs, unfortunately. So the steps cannot access the docker images that are running as service containers</p>
<p>Any ideas/suggestions? If I could only run just a custom command in the <code>backend</code> service… I tried entrypoint but it doesn’t yield the result I need. I created a post related to that: <a href="https://github.community/t/override-or-run-custom-command-in-an-action-service-container/232637" class="inline-onebox">Override or run custom command in an action service container</a></p>

**Comment by AssisrMatheus** [comment]
<aside class="quote group-Advisory-Council" data-username="airtower-luna" data-post="4" data-topic="232003">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The <code>docker</code> command talks to a service ( <code>dockerd</code> ) over a socket, and <code>dockerd</code> then runs the actual containers. The Actions runner makes sure that socket is available inside a job container</p>
</blockquote>
</aside>
<p>:point_up: looks like this isn’t being applied for my case? Since it can’t communicate with the running containers</p>

**Comment by airtower-luna** [comment]
<p>Can you share the link to a workflow run? That seems too complicated to figure out without the full context. :sweat_smile:</p>
<p>What I can say is that running an NPM oneshot command as entrypoint will make your service container terminate quickly.</p>

**Comment by AssisrMatheus** [comment]
<p>I don’t think I can share the workflow run, as it’s in a private repo. But the workflow comes from this file, you can see the docker install commands as steps on the test-chrome job, and it’s basically just the commands from the docker installation guide for debian:</p>
<pre><code class="lang-auto">name: Continuous Integration

on:
  workflow_dispatch:
  pull_request:
    branches: [dev]

jobs:
  install-and-build:
    timeout-minutes: 10
    runs-on: ubuntu-latest
    container: cypress/browsers:node14.15.0-chrome96-ff94
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          registry-url: 'https://npm.pkg.github.com'
      - name: Build and cache for cypress
        uses: cypress-io/github-action@v2
        with:
          runTests: false
          build: npm run build
      - name: Save build folder
        uses: actions/upload-artifact@v2
        with:
          name: .next
          if-no-files-found: error
          path: .next

  test-chrome:
    timeout-minutes: 15
    runs-on: ubuntu-latest
    container: cypress/browsers:node14.15.0-chrome96-ff94
    needs: install-and-build
    services:
      db:
        image: mysql
        ports:
          - 3306:3306
        env:
          MYSQL_ROOT_PASSWORD: root
        options: &gt;-
          --health-cmd "mysqladmin ping -h localhost --silent --wait=20"
          --health-interval 30s
          --health-timeout 25s
          --health-retries 10
          --health-start-period 5s
      backend:
        image: 'path/to/prebuilt/backend/image'
        ports:
          - 3001:3001
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          registry-url: 'https://npm.pkg.github.com'
      - name: Install docker
        run: |
          apt-get update &amp;&amp; apt-get install --no-install-recommends -y ca-certificates curl gnupg lsb-release sudo
          curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
          echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list &gt; /dev/null
          apt-get update &amp;&amp; apt-get install --no-install-recommends -y docker-ce docker-ce-cli containerd.io
      - name: Run seed
        run: |
          docker exec ${{ job.services.backend.id }} sh -c 'npm run seed'
          docker kill --signal=SIGHUP ${{ job.services.backend.id }}
      - name: Download the build folders
        uses: actions/download-artifact@v2
        with:
          name: .next
          path: .next
      - name: Run cypress with chrome
        uses: cypress-io/github-action@v2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          start: npm run test:serve
          wait-on: 'http://localhost:3001, http://localhost:3000'
          wait-on-timeout: 120
          browser: chrome
      - uses: actions/upload-artifact@v2
        if: failure()
        with:
          name: cypress-screenshots-chrome
          path: cypress/screenshots
      - uses: actions/upload-artifact@v2
        if: always()
        with:
          name: cypress-videos-chrome
          path: cypress/videos
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<p>I just tried accessing the Docker socket myself, and it worked flawlessly:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/airtower-luna/hello-github-actions/runs/5364011412?check_suite_focus=true">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/airtower-luna/hello-github-actions/runs/5364011412?check_suite_focus=true" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/5814919/181089971-03eff346-017e-4ab1-b716-454d977cf2be.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/7/374dd8eebf5063e7f545e944d956561c2b75ba2a_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/7/374dd8eebf5063e7f545e944d956561c2b75ba2a_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/7/374dd8eebf5063e7f545e944d956561c2b75ba2a.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/7/374dd8eebf5063e7f545e944d956561c2b75ba2a_2_10x10.png"></div>

<h3><a href="https://github.com/airtower-luna/hello-github-actions/runs/5364011412?check_suite_focus=true" target="_blank" rel="noopener">Inspect a job service from job container ·...</a></h3>

  <p>Contribute to airtower-luna/hello-github-actions development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Note that:</p>
<ul>
<li>The Docker socket is mounted (in the “Initialize containers” step, that’s done automatically by the runner).</li>
<li>I install Docker from the Debian repository, not <code>docker.com</code>, but that shouldn’t break anything.</li>
<li>Inspecting the service container by ID (similar to your <code>exec</code>) works.</li>
</ul>
<p>Combined with the error message you got saying that the backend container <em>is not running</em>, I’m pretty sure that the Docker socket is actually available, but there’s some kind of problem with the service container. Maybe it failed to start?</p>

**Comment by AssisrMatheus** [comment]
<p>Sorry for the delay, I was assigned to another task for a while.</p>
<p>But thank you so much for the help!!! In the end I think the issue was how I installed docker. I had other issues too, but that was after docker had been installed and I managed to connect to the service. <code>localhost</code> wasn’t working, but using the service names <code>db</code> and <code>backend</code> as their ip, just like as if it was a docker-compose, worked.</p>
